text/0000-vec-fallible-allocation.md

```diff
@@ -0,0 +1,341 @@
+- Feature Name: `vec_fallible_allocation`
+- Start Date: 2022-20-04
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+`Vec` has many methods that may allocate memory (to expand the underlying storage, or shrink it down). Currently each of these method rely on "infallible"
+allocation, that is any failure to allocate will call the global OOM handler, which will (typically) panic. Even if the global OOM handler does not panic, the
+return type of these method don't provide a way to indicate the failure.
+
+Currently `Vec` does have a `try_reserve` method that uses "fallible" allocation: if `try_reserve` attempts to allocate, and that allocation fails, then the
+return value of `try_reserve` will indicate that there was a failure, and the `Vec` is left unchanged (i.e., in a valid, usable state). We propose adding
+more of these `try_` methods to `Vec`, specifically for any method that can allocate.
+
+Unlike most RFCs, we are not suggesting that this proposal is the best among many alternatives (in fact we know that adding more `try_` methods is undesirable
+in the long term: see the "Drawbacks" section below), instead we are suggesting this as a way forward to unblock important work (see the "Motivations" section
+below) while we explore other alternatives. We have no plans to stabilize this feature in its current form and fully expect that these methods will be removed
+in the future.
+
+# Motivation
+[motivation]: #motivation
+
+The motivation for this change is well documented by other RFCs, such as the proposed [fallible_allocation feature](https://github.com/rust-lang/rfcs/pull/3140)
+and the accepted [fallible_collection_alloc feature](https://github.com/rust-lang/rfcs/blob/master/text/2116-alloc-me-maybe.md).
+
+As a brief summary, there are environments where dynamic allocation is required and failing to allocate is both expected and able to be handled. For example,
+in OS Kernel environments (such as [Linux](https://lore.kernel.org/lkml/CAHk-=wh_sNLoz84AUUzuqXEsYH35u=8HV3vK-jbRbJ_B-JjGrg@mail.gmail.com/)), some embedded
+systems, high-reliability systems (such as databases) and multi-user services (where a single request may fail without the entire service halting, such as a
+web server).
+
```

Comment:
```suggestion

Users in such environments may wish to make use of the `no_global_oom_handling` cfg added in <https://github.com/rust-lang/rust/pull/84266> in order to ban infallible allocation in their codebase's entire dependency graph. In such a situation they may still wish to use use the standard library's `Vec` type, but in combination with custom fallible allocation extension methods or a `FallibleVec` wrapper type provided by an external crate such as [fallible_collections](https://crates.io/crates/fallible_collections). However, once the `no_global_oom_handling` CFG is enabled, basic `Vec` methods like `push` are removed, making it difficult and unsafe to implement fallible equivalents outside of the standard library.

The additional methods proposed by this RFC would be enough for users in OOM-sensitive environments to make use of `no_global_oom_handling` without needing to fork or otherwise reimplement `Vec` externally. This is a sufficient stopgap solution for many use cases while a more complete fallible allocation solution is developed.

```

Per [my recent comment](https://github.com/rust-lang/rfcs/pull/3271#issuecomment-1279560444), and @ChrisDenton's suggestion to update the motivation section, here's my suggested update to the RFC text to capture the `no_global_oom_handling` use case. It looks like this was already mentioned in reference-level explanation, but hopefully adding it to the motivation section will also help clarify to readers how this RFC enables a use case that currently can't be accomplished.

Also sorry if the suggested wording isn't great, feel free to reword as needed if you agree this is worth adding!

Reply:
Thanks! This looks good to me. @dpaoliello do you want to accept it?